pypose/module/controller_parameters_tuner.py

```diff
@@ -0,0 +1,194 @@
+    r"""
+    Args:
+            learning_rate(float): gradient descent step size
+            penalty_coefficient(float): controller-effort penalty coefficient
+            device(string): on the cpu or cuda to perform the tuning process
+
+    This class is the general implementation of the controller parameters tuner based on
+    the sensitivity propagation.
+
+    Given a dynamic system, the system state transition can be defined as
+
+    .. math::
+            x_{i+1} = f(x_i, u_i) \tag{1}
+
+    .. math::
+            u_i = h(x_i, \hat{x_i}, \theta) \tag{2}
+
+    Where :math:`x_{i+1}` stands for the dynmamic system at time :math:`i+1`,
+    :math:`x_i` (:math:`x_i \in \mathbb{R}^n`) gives the dynamic system state at time :math:`i`,
+    :math:`u_i` (:math:`u_i \in \mathbb{R}^m`) is the controller-generated system inputs at time :math:`i`,
+    :math:`\theta` is the controller parameters.
+
+    Note:
+        For the parameter initial_state and parameters, please use 1d row vector to
+        represent these variables at the same time.
+
+
+
+    Given a dynamic system, a corresponding controller and reference states, assuming all
+    these components are differentiable, it is possbile to find the most suitable
+    parameters by constructing the loss function :math:`\textbf{L}` and computing the gradient of
+    :math:`\textbf{L}` with respect to controller parameters.
+
+
+
+    .. math::
+            \mathbf{L} = \sum_{i=1}^N ||x_i - \hat x_i||^2 + \sum_{i=0}^{N-1} \lambda|| u_i ||^2 \tag{3}
+
+    Where :math:`\mathbf{L}` is the Loss function, :math:`x_i` is the dynamic system state at time :math:`i`,
+    :math:`\hat x_i` is the desired dynamic system states predined by the user. :math:`u_i` is the dynamic system inputs
+    generated by the controller. Here it works as the control-effort penalty in the loss function. :math:`\lambda` is the
+    penalty coefficient.
+
+    And we can get the derivatives as following:
+
+    .. math::
+        \begin{align*}
+            \nabla L_{\theta} &= \sum_{i=1}^N \frac{\partial L}{\partial x_i} \frac{\partial x_i}{\partial \theta} + \sum_{i=0}^{N-1} \frac{\partial L}{\partial u_i} \frac{\partial u_i}{\partial \theta} \\
+            \frac{\partial x_{i+1}}{\partial \theta} &= (\frac{\partial x_{i+1}}{\partial x_i} + \frac{\partial x_{i+1}}{\partial u_i}\frac{\partial u_i}{\partial x_i})
+            \frac{\partial x_i}{\partial \theta}+\frac{\partial x_{i+1}}{\partial u_i}\frac{\partial u_i}{\partial \theta} \\
+            \frac{\partial u_i}{\partial \theta} &= \frac{\partial u_i}{\partial x_i} \frac{\partial x_i}{\partial \theta} + \frac{\partial u_i}{\partial \theta}
+        \end{align*} \tag{4}
+
+    To make sure the controller parameters can stay in the reasonable and safe sets, after updating the parameters, it's essential to project the new parameters into
+    the feasiable set. And the parameters can be updated using the following equation:
```

Comment:
The following update rule doesn't tell you are projecting the new parameters into the feasible set. It's just the normal gradient descent.